### PR TITLE
Fix capitalize Text style on IOS

### PR DESCRIPTION
**Pull request by MaeIg**
## Summary

On my project, I realized that capitalize style doesn't work with dates on IOS. I found [this issue](https://github.com/facebook/react-native/issues/32697) and tried to solve it.

(code example: https://snack.expo.dev/@maelg/capitalize-demo)
| Android | IOS | Web |
| ------------- | ------------- | ------------- |
| image | image | image |

As we can see, the behavior is different on IOS, Android and web: 
- **Android**: Capitalize the first letter of each word, unless it begins with a number. And put the rest in lowercase.
- **IOS**: Capitalize the first letter of each word, ~~unless it begins with a number~~. And put the rest in lowercase.
- **Web**: Capitalize the first letter of each word, unless it begins with a number. ~~And put the rest in lowercase.~~

This PR aims to unify behavior on Android and Ios. I am not changing the behavior which differs from the web because I don't know if it is desirable to align with the web.

## Changelog

[IOS] [Changed] - Don't capitalize the first letter of a word that is starting by a number

## Test Plan

I manually tested my changes on a POC app (same code: https://snack.expo.dev/@maelg/capitalize-demo) on react-native v0.66.4 and react-native main branch.

You can see the result here:
| Android | IOS |
| ------------- | ------------- |
| image | image |

~~I tried to use rn-tester but it was not taking my code. I think it is because fabric was enabled so it was using other code.
I tried to disable fabric but I was not able to build the app on my IOS simulator anymore:~~

On rn-tester:
<image src="https://user-images.githubusercontent.com/40902940/146457851-864b2962-5e9c-4c7e-83fd-7686e27cb996.png" width=50% height=50% />

### CODE REVIEW

**Comment by yungsters** [comment]
Thanks for the contribution! It is interesting that web preserves pre-existing capitalized letters. I agree that we do not need to consider that as part of this PR, but it is interesting to consider which behavior is desirable longer term.

**Comment by sota000** [comment]
Hi there! Can you try removing Pods folder, Podfile.lock, and build folder before running pod install with fabric_enabled off?

**Comment by MaeIg** [comment]
Thanks for your quick answers!

I managed to launch rn-tester without fabric thanks to you @sota000! I'll edit my first comment to add screenshots.
I found a crash on rn-tester when there are several spaces in a row. I just pushed a fix!

File: Libraries/Text/RCTTextAttributes.m [hunk 1]
```diff
@@ -237,6 +237,23 @@ - (UIColor *)effectiveBackgroundColor
   return effectiveBackgroundColor ?: [UIColor clearColor];
 }
 
+- (NSString *)capitalize:(NSString *)text
+{
+  NSArray *words = [text componentsSeparatedByString:@" "];
+  NSMutableArray *newWords = [[NSMutableArray alloc]init];
+  NSNumberFormatter *num = [[NSNumberFormatter alloc]init];
```
Comment (philIip) [inline Libraries/Text/RCTTextAttributes.m:244]: we prefer to use `new` if there's no initializer params

File: packages/rn-tester/js/examples/Text/TextExample.android.js [hunk 2]
```diff
@@ -798,6 +798,12 @@ class TextExample extends React.Component<{...}> {
           <Text style={{textTransform: 'capitalize'}}>
             This text should be CAPITALIZED.
           </Text>
+          <Text>
+            Capitalize a date:
+            <Text style={{textTransform: 'capitalize'}}>
+              the 9th of november, 1998
+            </Text>
+          </Text>
```
Comment (philIip) [inline packages/rn-tester/js/examples/Text/TextExample.android.js:806]: let's add one for 2 digit dates as well

File: Libraries/Text/RCTTextAttributes.m [hunk 3]
```diff
@@ -237,6 +237,23 @@ - (UIColor *)effectiveBackgroundColor
...
+  for (NSString *item in words) {
+    NSString *word; 
+    if ([item length] > 0 && [num numberFromString:[item substringWithRange:NSMakeRange(0, 1)]] == nil) {
+      word = [item capitalizedString];
+    } else {
+      word = [item lowercaseString];
+    }
+    [newWords addObject:word];
+  }
+  return [[newWords valueForKey:@"description"] componentsJoinedByString:@" "];
```
Comment (philIip) [inline Libraries/Text/RCTTextAttributes.m:254]: why aren't we just joining on `newWords`? NSString's description is just the string itself

File: Libraries/Text/RCTTextAttributes.m [hunk 4]
```diff
@@ -237,6 +237,23 @@ - (UIColor *)effectiveBackgroundColor
   return effectiveBackgroundColor ?: [UIColor clearColor];
 }
 
+- (NSString *)capitalize:(NSString *)text
```
Comment (philIip) [inline Libraries/Text/RCTTextAttributes.m:240]: can you make this a C function? i.e.

```
static NSString *CapitalizeText(NSString *text) {
  ...
}
```

this is more lightweight and the function does not require any state from this class.

**Review by philIip — CHANGES_REQUESTED** [review]

File: Libraries/Text/RCTTextAttributes.m (hunk 1 above)
Comment (MaeIg) [reply Libraries/Text/RCTTextAttributes.m]: Oh I didn't know this syntax. Done!

File: packages/rn-tester/js/examples/Text/TextExample.android.js (hunk 2 above)
Comment (MaeIg) [reply packages/rn-tester/js/examples/Text/TextExample.android.js:806]: Done! 
It still works:
image

File: Libraries/Text/RCTTextAttributes.m (hunk 3 above)
Comment (MaeIg) [reply Libraries/Text/RCTTextAttributes.m]: Yes you're right. Done!

File: Libraries/Text/RCTTextAttributes.m (hunk 4 above)
Comment (MaeIg) [reply Libraries/Text/RCTTextAttributes.m]: I think this is good. Tell me if it isn't!

As you can see with the code I wrote, I'm not familiar with Objective-C. I learned while I was making this PR 😁 So, thanks a lot for taking the time to review this PR, I learned a lot!

**Pull request closed without merging**